### PR TITLE
Add iOS support for numberOfLines prop on multiline TextInput

### PR DESCRIPTION
**Pull request by gilesvangruisen**
This adds functionality to allow multiline TextInput components to pass
the `numberOfLines` prop on iOS. If provided, input will not be allwoed
when that line count has been reached, but will still allow input
shorter than that line count.

This also adds the `numberOfLines` prop to the list of "only multiline" props in `TextInput.js`. Please let me know if this is errant, i.e. if Android `TextInput` can support both `multiline={false}` and `numberOfLines=…`. That seems somewhat illogical, so I figured this was a safe change.

**Motivation**:

I was trying to build an iOS text input that limited text to a certain line count (character count not an option), and realized there was no easy or good way to do this with the current iOS `TextInput`, and `numberOfLines` is only currently supported on the Android `TextInput`.

**Test plan**:

This PR adds an additional check to the `UITextViewDelegate`'s `- (BOOL)textView:shouldChangeTextInRange:replacementText:` method. It determines what the string _could_ be, and uses that new string to calculate what the size of the text view will be if it's accepted. Based on that new potential text view content size, it calculates the number of lines using the current font's `lineHeight`. If the new number of lines exceeds the `maximumNumberOfLines` on the text view's `textContainer` (set to the `numberOfLines` prop in `updateFrames`), the method will return `NO`, preventing the input from being committed to the text view. If it doesn't exceed the `numberOfLines`, the method will return `YES`, allowing the change to be made.

This kind of UI change is difficult to write a test for, but the calculation logic is sound, and I've tested it manually with a number of different textView sizes and `numberOfLines` prop values. It accounts for cases where the prop may be `undefined` or `0`, and in such cases will not limit the line count.

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
By analyzing the blame information on this pull request, we identified @JoelMarcey and @javache to be potential reviewers.

**Comment by facebook-github-bot-4** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by facebook-github-bot-2** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by gilesvangruisen** [comment]
Bump: @JoelMarcey and @javache should be a fairly straightforward review. Thanks!

**Comment by JoelMarcey** [comment]
Adding @bestander as the main reviewer -- please reassign as necessary.

**Comment by bestander** [comment]
I pass this along to @majak because he and @javache know more about iOS

File: Libraries/Text/RCTTextView.m [hunk 1]
```diff
@@ -336,6 +337,22 @@ - (void)setContentInset:(UIEdgeInsets)contentInset
 
 - (BOOL)textView:(RCTUITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
 {
+  if (textView.textContainer.maximumNumberOfLines > 0) {
```
Comment (javache) [inline Libraries/Text/RCTTextView.m]: This method is already pretty huge. Can you move this to where we evaluate maxLength?

**Comment by facebook-github-bot-9** [comment]
@gilesvangruisen updated the pull request.

**Comment by facebook-github-bot-9** [comment]
@gilesvangruisen updated the pull request - [view changes](https://github.com/facebook/react-native/pull/9011/files/cfe1a74ed6ff85cb4874454d65e4a09a634a084a..4e389e1b8d5478b6f78b7c1bd025048c04a4a58b)

**Comment by gilesvangruisen** [comment]
I'll have some time tomorrow to clean this up and fix a bug I recently found with this update, too.

**Comment by JoelMarcey** [comment]
cc @bestander

File: Libraries/Text/RCTTextView.m (hunk 1 above)
Comment (hramos) [inline Libraries/Text/RCTTextView.m]: Was this feedback addressed?

**Comment by lacker** [comment]
Looks like there is still some unaddressed feedback - @gilesvangruisen could you take a look? I see a non-responded-to comment when I click "files changed".

Also, could you add a test to ensure this behavior works? It seems risky to add a new parameter without any tests to ensure that it doesn't break.

**Comment by hramos** [comment]
Ping @gilesvangruisen do you still plan to ship this PR?

**Comment by shergin** [comment]
@gilesvangruisen Does this work identically on Android? Just stopping, no scrolling?

**Comment by hramos** [comment]
This pull request has been inactive for a while and the last comments are questions or there have been changes requested.

I'll close the pull request so it doesn't stay open indefinitely. If you plan to continue working on this please send a new pull request.

**Pull request closed without merging**